File: example/calc.js
function inherits(child, parent) {
  child.prototype = Object.create(parent.prototype, {
    constructor: {
      value: child,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
}


function E() {}
function EInt(x) { this.x = x; } inherits(EInt, E);
function EAdd(x, y) { this.x = x; this.y = y; } inherits(EAdd, E);
function EMul(x, y) { this.x = x; this.y = y; } inherits(EMul, E);

EInt.prototype.eval = function() { return this.x; };
EAdd.prototype.eval = function() { return this.x.eval() + this.y.eval(); };
EMul.prototype.eval = function() { return this.x.eval() * this.y.eval(); };

function main() {

  var i = new EInt(1);
  console.log("eval 1 = " + i.eval());

  var add = new EAdd(new EInt(1), new EInt(2));
  console.log("eval 1 + 2 = " + add.eval());

  var mul = new EMul(new EAdd(new EInt(1), new EInt(2)), new EInt(111));
  console.log("eval (1 + 2) * 111 = " + mul.eval());

}

main();

File: example/calc.rs
extern crate std;

struct EInt {x:int}
struct EAdd {x:Box<E>,y:Box<E>}
struct EMul {x:Box<E>,y:Box<E>}

trait E         { fn eval(&self) -> int; }
impl E for EInt { fn eval(&self) -> int { self.x } }
impl E for EAdd { fn eval(&self) -> int { self.x.eval() + self.y.eval() } }
impl E for EMul { fn eval(&self) -> int { self.x.eval() * self.y.eval() } }

fn main() {

  let i = EInt{x:1};
  println!("eval 1 = {}", i.eval());

  let add = EAdd{x:box EInt{x:1},y:box EInt{x:2}};
  println!("eval 1 + 2 = {}", add.eval());

  let mul = EMul{x:box EAdd{x:box EInt{x:1},y:box EInt{x:2}},y:box EInt{x:111}};
  println!("eval (1 + 2) * 111 = {}", mul.eval());
}
